sound/soc/sof/intel/hda.c

```diff
@@ -1176,8 +1176,12 @@ int hda_dsp_remove(struct snd_sof_dev *sdev)
 				SOF_HDA_INT_CTRL_EN | SOF_HDA_INT_GLOBAL_EN, 0);
 
 	/* disable cores */
-	if (chip)
-		hda_dsp_core_reset_power_down(sdev, chip->host_managed_cores_mask);
+	if (chip) {
+		if (chip->hw_ip_version > SOF_INTEL_CAVS_2_5)
+			mtl_dsp_power_down_dsp_subsystem(sdev);
+		else
+			hda_dsp_core_reset_power_down(sdev, chip->host_managed_cores_mask);
+	}
 
```

Comment:
suggest another way. First change it to

````c
if (chip->host_managed_cores_mask == SOF_DSP_PRIMARY_CORE) // ON TGL+ it will be true
           snd_sof_dsp_core_put(sdev, SOF_DSP_PRIMARY_CORE)
else
          hda_dsp_core_reset_power_down(sdev, chip->host_managed_cores_mask);
````
in mtl.c, you can implement mtl_dsp_core_put like tgl_dsp_core_put 
````c
static int tgl_dsp_core_put(struct snd_sof_dev *sdev, int core)
{
	const struct sof_ipc_pm_ops *pm_ops = sdev->ipc->ops->pm;

	/* power down primary core and return */
	if (core == SOF_DSP_PRIMARY_CORE)
		return hda_dsp_core_reset_power_down(sdev, BIT(core));

	if (pm_ops->set_core_state)
		return pm_ops->set_core_state(sdev, core, false);

	return 0;
}
````

Reply:
go further:  there is not core_put for platforms before tgl, so you can also implement core_put for each platform to do this. Then we can just call core_put here.

Reply:
@RanderWang we cannot implement core_put for the platforms before TGL and there's a specific reason for it that the FW doesnt support disabling secondary cores. We made a decision particularly keeping in mind that we are not going to fix multi-core for the older platforms.